File: I-InterfaceSegregation/Good_IWorker.cpp
// ✅ GOOD — Follows Interface Segregation Principle
// Split the fat interface into smaller, focused ones.

#include <iostream>

// Every worker can work
class IWorkable {
public:
    virtual void Work() = 0;
    virtual ~IWorkable() = default;
};

// Only biological workers need to eat
class IEatable {
public:
    virtual void Eat() = 0;
    virtual ~IEatable() = default;
};

// Human implements BOTH interfaces
class HumanWorker : public IWorkable, public IEatable {
public:
    void Work() override { std::cout << "Human working...\n"; }
    void Eat()  override { std::cout << "Human eating lunch...\n"; }
};

// Robot only implements what it needs ✅
class RobotWorker : public IWorkable {
public:
    void Work() override { std::cout << "Robot working 24/7...\n"; }
    // No Eat() — no forced exception ✅
};

// Usage:
// HumanWorker h; RobotWorker r;
// h.Work(); h.Eat();
// r.Work(); // fine ✅

File: I-InterfaceSegregation/Good_IWorker.cs
// ✅ GOOD — Follows Interface Segregation Principle
// Split the fat interface into smaller, focused ones.
// Each class only implements what it actually needs.

// Every worker can work
public interface IWorkable
{
    void Work();
}

// Only biological workers need to eat
public interface IEatable
{
    void Eat();
}

// Optional: a base marker if needed
public interface IWorker : IWorkable { }

// Human implements BOTH interfaces
public class HumanWorker : IWorkable, IEatable
{
    public void Work() => Console.WriteLine("Human working...");
    public void Eat()  => Console.WriteLine("Human eating lunch...");
}

// Robot only implements what it needs ✅
public class RobotWorker : IWorkable
{
    public void Work() => Console.WriteLine("Robot working 24/7...");
    // No Eat() — no forced NotImplementedException ✅
}

// Usage:
// IWorkable[] workers = { new HumanWorker(), new RobotWorker() };
// foreach (var w in workers) w.Work(); // Both work fine ✅
